relayer/listener.go

```diff
@@ -300,38 +232,21 @@ func (lstnr *Listener) ProcessLogs(ctx context.Context) error {
 					)
 					continue
 				}
-				if msgInfo == nil {
-					lstnr.logger.Debug(
-						"Application relayer not found. Skipping message relay.",
-						zap.String("blockchainID", lstnr.sourceBlockchain.GetBlockchainID().String()),
-					)
+				if appRelayer == nil {
+					lstnr.logger.Debug("Application relayer not found. Skipping message relay")
 					continue
 				}
-				msgsInfo = append(msgsInfo, msgInfo)
-				expectedMessages[msgInfo.applicationRelayer.relayerID]++
+				messageHandlers[appRelayer.relayerID.ID] = append(messageHandlers[appRelayer.relayerID.ID], handler)
 			}
+			// Initiate message relay of all registered messages
 			for _, appRelayer := range lstnr.applicationRelayers {
-				// Prepare the each application relayer's database key with the number
-				// of expected messages. If no messages are found in the above loop, then
-				// totalMessages will be 0
-				totalMessages := expectedMessages[appRelayer.relayerID]
-				appRelayer.checkpointManager.PrepareHeight(block.BlockNumber, totalMessages)
-			}
-			for _, msgInfo := range msgsInfo {
-				lstnr.logger.Info(
-					"Relaying message",
-					zap.String("sourceBlockchainID", lstnr.sourceBlockchain.GetBlockchainID().String()),
-					zap.String("warpMessageID", msgInfo.unsignedMessage.ID().String()),
-				)
-				err := lstnr.dispatchToApplicationRelayer(msgInfo, block.BlockNumber)
-				if err != nil {
-					lstnr.logger.Error(
-						"Error relaying message",
-						zap.String("sourceBlockchainID", lstnr.sourceBlockchain.GetBlockchainID().String()),
-						zap.Error(err),
-					)
-					continue
-				}
+				// Dispatch all messages in the block to the appropriate application relayer.
+				// An empty slice is still a valid argument to ProcessHeight; in this case the height is immediately committed.
+				handlers := messageHandlers[appRelayer.relayerID.ID]
+
+				// Process the height async. This is safe because the ApplicationRelayer maintains the threadsafe
+				// invariant that heights are committed to the database one at a time, in order, with no gaps.
+				go appRelayer.ProcessHeight(block.BlockNumber, handlers)
```

Comment:
The first version of this PR called `appRelayer.ProcessHeight` synchronously. `ProcessHeight` manages the `ProcessMessage` goroutines, and blocks until all message workers have finished. As a result, the call here would block for each application relayer that the `Listener` manages. 

The fix is to call `ProcessMessage` async. I also had to add a lock acquisition to `checkpoint.StageProcessedHeight`, since there is no longer a guarantee that blocks finish processing in order.